Handreiking API Design/Handreiking APi Design.md

```diff
@@ -16,24 +18,31 @@ Dit document heeft een tweeledig doel.
 
 **Motivatie en dominante domeinen**
 
-Bij relaties tussen twee domeinen wil je vanuit elk domein de gerelateerde
-gegevens uit het andere domein kunnen opvragen/inkijken. Eén van de domeinen
-wordt als dominant/primair domein beschouwt. Soms is heel duidelijk welke dat
+Bij relaties tussen twee domeinen wil je vanuit beide domeinen de gerelateerde
+gegevens uit het andere domein kunnen opvragen/inkijken. Daarvoor wordt in elk domein een relatie-entiteit gedefinieerd die in feite de representatie van een entiteit in een ander domein is. Zo'n relatie-entiteit kan vervolgens gebruikt worden om aan andere entiteiten binnen het domein te koppelen. Dit is noodzakelijk omdat de relatie tussen de entiteit in het ene domein en de entiteit in het andere domein slechts op logisch niveau kan worden gelegd. Zo'n relatie wordt nooit op technisch niveau gecreëerd. 
+
+In het voorbeeld hieronder zie je dat in het 'Zaken API' domein de relatie entiteit 'zaakverzoek' is gecreëerd wat in feite een representatie is van de entiteit 'verzoek' uit het 'Klantinteracties API' domein. In het 'Klantinteracties API' domein is juist de relatie-entiteit 'objectverzoek' gecreëerd wat een representatie is van de entiteit 'zaak' uit het 'Zaken API' domein.
```

Comment:
zaakverzoek is een koppeltabel, dus met alleen maar twee verwijzingen (identifiers, uri's, identificaties). Het is geen representatie in de zin van een kopie van het verzoek opbject. 

Misschien bedoelde je dat ook maar bij representatie denk ik toch aan een klasse met attributen, niet aan een koppeling.

Reply:
Nee, ik bedoelde eigenlijk een soort van placeholder waarbij je niet alle attributen herhaald. Dat willen we immers juist niet. Ik realiseer me echter inmiddels dat het karakter van de relatie-entiteiten toch even anders is dan ik dacht. Het is geen placeholder maar juist een koppeltabel. Ik ga dit stuk dus herschrijven en mogelijk heeft dat ook gevolgen voor de rest van dit document.

Reply:
@JohanBoer, ik denk dat jij je commentaar in de conversation view hebt toegevoegd. Daardoor heb je denk ik niet gezien dat ik dit deel van het document al herschreven heb n.a.v. de opmerking van Michiel. Ik negeer je opmerkingen dus maar.

Reply:
> We moeten hier uitkijken dat we niet 1 op 1 de oplossing van de developers overnemen

Zie niet in waarom dat slecht is :)

> Bv. Als die relatie helemaal geen eigenschappen heeft kan je ervoor kiezen om de koppeltabellen die in beide domeinen worden vastgelegd helemaal niet op te nemen in de API, maar gewoon de gerelateerde objecten mee te geven in een array.

Technisch kan dat. Sterker nog, dat was onze eerste oplossing en hij is visueel ook het mooist maar precies onder invloed van onze "developer mindset", waarmee we in de praktijk de API's ook echt bouwen, hebben we deze oplossing ook snel weer afgeschaft.

Deze oplossing (arrays op de gerelateerde resource) heeft als groot nadeel dat je de gerelateerde resource (bijv. ``Zaak``) moet **updaten** in plaats van een relatie resource moet **aanmaken** of **verwijderen** (``ZaakVerzoek``).

Het updaten van arrays is bijna altijd verwarrend in de praktijk\*. Het kan zeker maar omdat je een resource moet bijwerken dat eigenlijk niets met de relatie zelf te maken heeft, hebben we gekozen voor een aparte relatie resource die je aanmaakt of verwijdert zonder dat je per ongeluk iets verkeerd bijwerkt op de gerelateerde resource (bijv. van de ``Zaak`` zelf dus).

---

\* Update van een array is verwarrend omdat er direct allerlei vragen ontstaan. Stel een resource ``foobar`` met de inhoud ``{ "naam": "test", "array": [1, 2] }``

* Wat doet PUT ``{ "naam": "test", "array": [] }`` -- Laat de array ongemoeid of overschrijft deze met een lege lijst?
* Wat doet PUT ``{ "naam": "test" }`` -- Laat de array ongemoeid of overschrijft deze met een lege lijst?
* Wat doet PUT ``{ "naam": "test", "array": [3, 4] }`` -- Overschrijft de hele array of voegt 3 en 4 toe?
* Wat doet PUT ``{ "naam": "test", "array": [1, 4] }`` -- Overschrijft de hele array of voegt alleen 4 toe?
* En je kan natuurlijk per ongeluk de naam (inhoudelijk) aanpassen terwijl je alleen maar een relatie wil aanpassen.

Overigens kan je op alle vragen antwoord geven en dit vastleggen ergens maar het feit dat de vragen zo makkelijk ontstaan is al onwenselijk.